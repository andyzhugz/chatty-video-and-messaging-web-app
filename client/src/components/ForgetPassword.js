import React, { useState } from 'react';
import { useHistory } from 'react-router-dom';
import { changePassword } from '../javascripts/authRequests'
import '../stylesheets/Profile.css';

const ForgetPassword = () => {

  const [error, setError] = useState(false);
  let history = useHistory();
  
  const validation = (password) => {
    let pattern = new RegExp(/^[0-9a-zA-Z]{8,}$/);
    if (!pattern.test(password)) {
        const msg = "At least 8 alphanumeric characters for password";
        setError(msg); 
        return false;
      }
    return true;
  };

  const handleChangePassword = (event) => {
    event.preventDefault();
    let email = document.getElementById('userName').value;
    let newPassword = document.getElementById('newPassword').value;
    if (validation(newPassword)) {
      changePassword(email, newPassword).then((res) => {
        if (res.status === 200) {
          localStorage.clear();
          history.push("/login");
        }
        else {
          setError("Failed to change password, non-existed user");
        }
      }).catch(() => {
        setError("Failed to change password, non-existed user");
      });
    }
  }

  return (
    <div>
      <h1 align="center">change Password</h1>
      <div className="mainProfile">
        <form className="formLogin">
          <input
            id="userName"
            className="newPassword"
            type="text"
            align="center"
            placeholder="username"
          />
          <input
            id="newPassword"
            className="newPassword"
            type="password"
            align="center"
            placeholder="new password"
          />
          <button id="resetBtn" className="resetBtn" align="center" onClick={handleChangePassword}>
            <b>Change Password</b>
          </button>
        </form>
        {error && <><div id="error" style={{ color: 'red', fontSize: 12, textAlign: "center" }}>{error}</div><br /></>}<br />
      </div>
    </div>
  );
};

export default ForgetPassword;
