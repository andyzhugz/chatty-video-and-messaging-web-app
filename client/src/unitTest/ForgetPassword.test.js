import { validation, ForgetPassword } from '../components/ForgetPassword.js'
import { BrowserRouter as Router } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import ReactDOM from 'react-dom';

const renderer = require('react-test-renderer');

// snapshot testings
describe('Test snapshot', () => {
    test('Test Login', () => {
      const component = renderer.create(<Router><ForgetPassword /></Router>);
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

describe('Independent function tests', () => {
    test('validation function true', () => {
        expect(validation('123456789')).toBe(true);
    });

    test('validation function false', () => {
        expect(validation('1')).toBe(false);
    });
  });
  
    describe('Test functonal component functions', () => {

      let emit;
      let container;
      beforeAll(() => {
        ({ emit } = window._virtualConsole);
      });

      beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        window._virtualConsole.emit = jest.fn();
      });
    
      afterEach(() => {
        document.body.removeChild(container);
        container = null;
      });
    
      test('Test function inside ForgetPassword component', async() => {
        act(() => {
          ReactDOM.render(<Router><ForgetPassword /></Router>, container);
        });
        const loginBtn = container.querySelector('#resetBtn');
        await act(async () => {
          fetch = jest.fn(() => Promise.resolve({ json: () => ({ message: 'success', data: [{ username: 'tester', password: 'tester' }] }) }));
          loginBtn.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        });
        const userName = container.querySelector('#userName');
        expect(userName.value).toBe("");
      });
});