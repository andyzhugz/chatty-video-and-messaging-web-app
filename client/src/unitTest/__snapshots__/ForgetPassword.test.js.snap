// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot Test Login 1`] = `
<div>
  <h1
    align="center"
  >
    change Password
  </h1>
  <div
    className="mainProfile"
  >
    <form
      className="formLogin"
    >
      <input
        align="center"
        className="newPassword"
        id="userName"
        placeholder="username"
        type="text"
      />
      <input
        align="center"
        className="newPassword"
        id="newPassword"
        placeholder="new password"
        type="password"
      />
      <button
        align="center"
        className="resetBtn"
        id="resetBtn"
        onClick={[Function]}
      >
        <b>
          Change Password
        </b>
      </button>
    </form>
    <br />
  </div>
</div>
`;
